subprojects/plugins/src/main/java/org/gradle/api/tasks/GroovyRuntime.java

```diff
@@ -91,15 +97,31 @@ public FileCollection createDelegate() {
                     throw new GradleException(String.format("Cannot infer Groovy class path because no Groovy Jar was found on class path: %s", Iterables.toString(classpath)));
                 }
 
-                if (groovyJar.isGroovyAll() || groovyJar.getVersion().getMajor() == 3) {
+                if (groovyJar.isGroovyAll()) {
                     return project.getLayout().files(groovyJar.getFile());
                 }
 
+                VersionNumber groovyVersion = groovyJar.getVersion();
+                // Groovy 3 does not have groovy-all yet we may have the required pieces on classpath via localGroovy()
+                if (groovyVersion.getMajor() == 3) {
+                    Set<String> groovyLibNames = GROOVY3_LIBS.stream()
+                        .map(libName -> libName + "-" + groovyVersion.toString() + ".jar")
+                        .collect(Collectors.toSet());
+                    List<File> groovyClasspath = new ArrayList<>();
+                    for (File file : classpath) {
+                        if (groovyLibNames.contains(file.getName())) {
+                            groovyClasspath.add(file);
+                        }
+                    }
```

Comment:
This could be one stream by using filter and collect into `groovyClasspath`

Reply:
I am struggling to come up with a nice way to simplify this without having to put one loop inside another (with streams).
Roughly: `classpath.stream().filter(f -> for each lib in GROOVY3_LIBS, if f.name == "$lib-$version.jar").collect();`
That loop in the middle would be the same as doing current
```
groovyJarNames = GROOVY3_LIBS.stream()
                    .map(libName -> libName + "-" + groovyVersion + ".jar")
                    .collect(Collectors.toSet());
```
but it would then be done once per every file in the classpath.